monix-eval/shared/src/main/scala/monix/eval/Task.scala

```diff
@@ -783,54 +801,57 @@ object Task extends TaskInstances {
     *
     * Although the effects are unordered, we ensure the order of results
     * matches the order of the input sequence. Also see [[gatherUnordered]].
-    *
-    * Alias for [[zipList]].
     */
-  def gather[A](in: Seq[Task[A]]): Task[List[A]] =
-    zipList(in)
+  def gather[A, M[X] <: TraversableOnce[X]](in: M[Task[A]])
+    (implicit cbf: CanBuildFrom[M[Task[A]], A, M[A]]): Task[M[A]] = {
+    val init = evalAlways(cbf(in))
+    val r = in.foldLeft(init)((acc,elem) => Task.mapBoth(acc,elem)(_ += _))
+    r.map(_.result())
+  }
 
-  /** Nondeterministically gather results from the given sequence of tasks
-    * to a sequence, without keeping the original ordering of results.
+  /** Nondeterministically gather results from the given collection of tasks,
+    * without keeping the original ordering of results.
     *
     * This function is similar to [[gather]], but neither the effects nor the
     * results will be ordered. Useful when you don't need ordering because it
     * can be more efficient than `gather`.
     */
-  def gatherUnordered[A](in: Seq[Task[A]]): Task[Seq[A]] =
+  def gatherUnordered[A, M[X] <: TraversableOnce[X]](in: M[Task[A]])
+    (implicit cbf: CanBuildFrom[M[Task[A]], A, M[A]]): Task[M[A]] =
     Async { (scheduler, conn, finalCallback) =>
-      val atom = Atomic(Vector.empty[A])
-      val expected = in.length
+      var remaining = in.size
```

Comment:
Dude, I think here we might have a problem. If we have a true `TraversableOnce`, it means we can traverse it only once, being the equivalent of an `Iterator`. And this signature allows it.

One proposal would be to just convert it into a `Seq` before using it, since a `Seq` can be traversed multiple times and the normal usecase will be for people to pass us Seqs anyway.

Another possible but more complicated solution that does traverse that only once would be:

1. we initialize `var remaining = 1` (one extra unit, because we don't want to trigger the final callback too early)
2. we place the while loop inside a `lock.synchronize` as well
3. we do `remaining += 1` on each cycle, before running the task
4. at the end of the loop, but before we leave the `synchronize`, we do `remaining -= 1`, but at this point the result might be zero already, so we need to trigger the `finalCallback` doing the `if (remaing == 0)` check just like in `onSuccess`